platform/default/src/mbgl/storage/offline_database.cpp

```diff
@@ -403,147 +469,224 @@ std::optional<std::pair<Response, uint64_t>> OfflineDatabase::getResource(const
     return std::make_pair(response, size);
 }
 
-std::optional<int64_t> OfflineDatabase::hasResource(const Resource& resource) {
-    mapbox::sqlite::Query query{getStatement("SELECT length(data) FROM resources WHERE url = ?")};
-    query.bind(1, resource.url);
-    if (!query.run()) {
-        return std::nullopt;
+std::optional<int64_t> OfflineDatabase::extractResourceDataSize(const Resource& resource, const char* sql) {
+    std::optional<int64_t> selectResourcesResult;
+    mapbox::sqlite::Query selectResourcesQuery(getStatement(sql));
+
+    selectResourcesQuery.bind(1, resource.url);
+
+    if (selectResourcesQuery.run()) {
+        selectResourcesResult = selectResourcesQuery.get<std::optional<int64_t>>(0);
+    } else {
+        selectResourcesResult = std::nullopt;
     }
+    return selectResourcesResult;
+}
+
+std::optional<int64_t> OfflineDatabase::hasResource(const Resource& resource) {
+    // First, try to find the resource in the 'resources' table
+
+    std::optional<int64_t> selectResourcesResult = extractResourceDataSize(
+        resource, "SELECT length(data) FROM resources WHERE url = ?1");
 
-    return query.get<std::optional<int64_t>>(0);
+    if (selectResourcesResult) {
+        return selectResourcesResult;
+    } else {
+        // If not found in 'resources', try the 'ambient_resources' table
+
+        std::optional<int64_t> selectAmbientResourcesResult = extractResourceDataSize(
+            resource, "SELECT length(data) FROM ambient_resources WHERE url = ?1");
+
+        if (selectAmbientResourcesResult) {
+            mapbox::sqlite::Query insertQuery(
+                getStatement("INSERT INTO resources (url, kind, expires, modified, etag, data, compressed, accessed, "
+                             "must_revalidate) "
+                             "SELECT url, kind, expires, modified, etag, data, compressed, accessed, must_revalidate "
+                             "FROM ambient_resources "
+                             "WHERE url = ?1"));
+            insertQuery.bind(1, resource.url);
+            insertQuery.run();
+
+            mapbox::sqlite::Query deleteQuery(getStatement("DELETE FROM ambient_resources WHERE url = ?1"));
+            deleteQuery.bind(1, resource.url);
+            deleteQuery.run();
+        }
+        return selectResourcesResult;
+    }
 }
 
-bool OfflineDatabase::putResource(const Resource& resource,
-                                  const Response& response,
-                                  const std::string& data,
-                                  bool compressed) {
+bool OfflineDatabase::putResource(
```

Comment:
There is too much code duplication in this function and the `putTile` function below.

Instead of adding a `bool` I would for example define an `enum class` that can be converted to a `string` to get the related database name using another utility function.

Reply:
We agree with you about the code duplication issue, but we encountered difficulties in resolving it :-(

Initially, we attempted to minimize code duplication in methods like getTile(), getResource(), putTile(), putResource(), etc., by using lambdas and c_str() to concatenate the table name in the query. However, this approach led to memory issues (EXC_BAD_ACCESS) and erratic behavior, particularly errors during the insertion of tiles/resources, for example : 
_[DEBUG] {abort at 23 in [INSERT INTO ambient_tiles (url_template, pixel_ratio, x, y, z, modified, must_revalidate, etag, expires, accessed, data, compressed) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)]: NOT NULL constraint failed: ambient_tiles.compressed}[Database](1299): %s_

Consequently, we had to revert our changes, even though it resulted in increased code duplication.

You can find here an old version of the code here with our trials for getTile() and getResource() method: https://github.com/geolives/maplibre-native/commit/cb8cf110e3675b022b99b0424d1db44a7b1a86d3

As our expertise in C++ is not as strong as it is in other higher-level languages, we welcome any assistance from the community to help resolve this issue of code duplication without causing adverse effects :-)

Thanks in advance.